code/iaas/model/src/main/java/io/cattle/platform/core/dao/impl/VolumeDaoImpl.java

```diff
@@ -96,4 +98,20 @@ public void createVolumeInStoragePool(Map<String, Object> volumeData, String vol
 
         return volumes;
     }
+
+    @Override
+    public Set<? extends Volume> findNonremovedVolumesWithNoOtherMounts(long instanceId) {
+        List<VolumeRecord> vols = create()
+            .selectDistinct(VOLUME.fields())
+            .from(MOUNT)
+            .join(VOLUME)
+                .on(VOLUME.ID.eq(MOUNT.VOLUME_ID).and(VOLUME.REMOVED.isNull()))
+            .where(MOUNT.VOLUME_ID.in(
```

Comment:
We had a similar subselect in @alena1108's code (can't remember where), but it killed mysql.  It was way faster to actually do the select(), get the result and then pass the result as an are to `in`.  This approach will work fine as long as you know the data set from the first select will typically be smallish (less than 100).